### PR TITLE
Optimisation inhibition

### PR DESCRIPTION
**Pull request by Praetonus**
[Rendered](https://github.com/Praetonus/rfcs/blob/opt/text/0000-optimisation-inhibition.md)

### CODE REVIEW

**Comment by jemc** [comment]
Are there examples of other languages that provide features like this?  Rust, maybe?  If so, what do they call them?  I'd be more likely to be in favor of this RFC if there is some precedent for it that users might be familiar with and we could use similar language to describe the features.

**Comment by Praetonus** [comment]
Rust has direct support of inline assembly so it doesn't need compiler support for this, as in C. That said, it looks like the standard benchmark library of Rust has a function named `black_box` wrapping the same assembly. Google's benchmark library (C++) uses `DoNotOptimize`. These two functions have the same goal as `escape` in the RFC: keeping a computation producing an arbitrary result untouched by the optimiser.

**Comment by jemc** [comment]
I must admit I still don't have a good idea of what `clobber` is doing.

**Comment by Praetonus** [comment]
`clobber` is basically a "touch all memory" function. This is useful with `escape`d pointers. `escape` is saying "I'm reading this input and doing something in memory" and `clobber` is saying "I'm doing something in memory". From the compiler standpoint, it could mean that `escape` stored the pointer in some global state and then `clobber` read the pointed value from the global state (but nothing happened in reality).

To sum it up, any write to an `escape`d memory location before a call to `clobber` must stay in the program.

**Comment by jemc** [comment]
I'm not sure about the name `Optimiser`, given that what this primitive actually does is _prevent_ optimization - the name definitely leads you to believe otherwise.

**Comment by Praetonus** [comment]
What do you think about `DoNotOptimise` for the primitive? It's not a noun, I don't know if that's problematic. For the functions, I'm thinking `apply` (for the sugar) and `observe` (because it's observing potential side effects) for `escape` and `clobber` respectively. I'll update the RFC if there are no objections against these names.

**Comment by jemc** [comment]
That name is definitely clearer, and I can't think of anything better, so I don't think I have any objections to it.  But we can hear what others think.

**Comment by Praetonus** [comment]
Since there hasn't been any new elements, I'd like to push this RFC to the final comment period.

**Comment by SeanTAllen** [comment]
@Praetonus can you send a short email to pony+dev with a link to this RFC, a short description and a request for any final comments? Thanks!

**Comment by Praetonus** [comment]
Sorry, I had no time to do it during the last days. I see you've sent the mail, thanks.

**Comment by SeanTAllen** [comment]
@Praetonus not a problem.

**Comment by sylvanc** [comment]
I like `DoNotOptimise` as it results in code like:

``` pony
actor Main
  new create(env: Env) =>
    var x = U32(0)
    DoNotOptimise[U32](x)
    x = 42
    DoNotOptimise.observe()
```

That reads nicely to me.

**Comment by SeanTAllen** [comment]
@Praetonus are you ready for this to move to an up/down vote at the next sync call?

**Comment by Praetonus** [comment]
@SeanTAllen Yes.
